toffy/streak_detection.py

```diff
@@ -0,0 +1,429 @@
+
+    Args:
+        shape (tuple): The shape of the image / fov.
+        streak_channel (str): The specific channel name used to create the masks.
+        corrected_dir (Path): The directory used to save the corrected tiffs and data in.
+        streak_mask (np.ndarray): The first binary mask indicating candidate streaks.
+        streak_df (pd.DataFrame): A dataframe, containing the location, area, and eccentricity
+        of each streak.
+        filtered_streak_mask (np.ndarray): A binary mask with out the false streaks.
+        filtered_streak_df (pd.DataFrame): A subset of the `streak_df` containing location, area
+        and eccentricity values of the filtered streaks.
+        boxed_streaks (np.ndarray): An optional binary mask containing an outline for each
+        filtered streaks.
+        corrected_streak_mask (np.ndarray): An optional binary mask containing the lines used for
+        correcting the streaks.
+    """
+
+    shape: tuple = None
+    streak_channel: str = None
+    corrected_dir: Path = None
+    streak_mask: np.ndarray = None
+    streak_df: pd.DataFrame = None
+    filtered_streak_mask: np.ndarray = None
+    filtered_streak_df: pd.DataFrame = None
+    boxed_streaks: np.ndarray = None
+    corrected_streak_mask: np.ndarray = None
+
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+    def save_data(self, name: Optional[str] = None, all_data: bool = True):
+        """Given a field name, it saves the data as a tiff file if it's a Numpy array, and a csv
+        if it is a Pandas DataFrame.
+
+        Args:
+            name (str): The field of the DataClass to be saved. Options include:
+                `streak_mask`,
+                `streak_df`,
+                `filtered_streak_mask`,
+                `filtered_streak_df`,
+                `boxed_streaks`,
+                `corrected_streak_mask`,
+            all_data (bool): A flag for saving all the masks and dataframes. Defaults to `True`.
+        """
+        fields = [
+            "streak_mask",
+            "streak_df",
+            "filtered_streak_mask",
+            "filtered_streak_df",
+            "boxed_streaks",
+            "corrected_streak_mask",
+        ]
+        if all_data:
+            for field in fields:
+                self._save(name=field)
+        elif name not in fields:
+            raise ValueError(
+                "Invalid field to save. Options include:\
+                \n streak_mask\
+                \n streak_df\
+                \n filtered_streak_mask\
+                \n filtered_streak_df\
+                \n boxed_streaks\
+                \n corrected_streak_mask"
+            )
+        elif name is not None:
+            self._save(name)
+        else:
+            raise ValueError("Please enter a field name, or set `all_data` to True.")
+
+    def _save_dir(self, data_dir: Path, name: str, ext: str):
```

Comment:
Function names should describe what they're doing. This name makes it sound like `save` is the verb, when in fact what this function actually does is `get_save_dir`